File: components/query_tiles/android/java/src/org/chromium/components/query_tiles/bridges/TileConversionBridge.java
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.query_tiles.bridges;

import androidx.annotation.Nullable;

import org.jni_zero.CalledByNative;
import org.jni_zero.JNINamespace;
import org.jni_zero.JniType;

import org.chromium.components.query_tiles.QueryTile;
import org.chromium.url.GURL;

import java.util.ArrayList;
import java.util.List;

/** Bridge to the native query tile service for the given {@link Profile}. */
@JNINamespace("query_tiles")
public class TileConversionBridge {
    @CalledByNative
    private static List<QueryTile> createList() {
        return new ArrayList<>();
    }

    @CalledByNative
    private static QueryTile createTileAndMaybeAddToList(
            @Nullable List<QueryTile> list,
            @JniType("std::string") String tileId,
            @JniType("std::string") String displayTitle,
            @JniType("std::string") String accessibilityText,
            @JniType("std::string") String queryText,
            @JniType("std::vector<const GURL*>") GURL[] urls,
            @JniType("std::vector<std::string>") String[] searchParams,
            List<QueryTile> children) {
        QueryTile tile =
                new QueryTile(
                        tileId,
                        displayTitle,
                        accessibilityText,
                        queryText,
                        urls,
                        searchParams,
                        children);
        if (list != null) list.add(tile);
        return tile;
    }
}

File: components/query_tiles/android/tile_conversion_bridge.cc
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/query_tiles/android/tile_conversion_bridge.h"

#include <memory>
#include <string>

#include "base/android/jni_string.h"
#include "url/android/gurl_android.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "components/query_tiles/jni_headers/TileConversionBridge_jni.h"

namespace query_tiles {

ScopedJavaLocalRef<jobject> CreateJavaTileAndMaybeAddToList(
    JNIEnv* env,
    ScopedJavaLocalRef<jobject> jlist,
    const Tile& tile) {
  ScopedJavaLocalRef<jobject> jchildren =
      Java_TileConversionBridge_createList(env);

  for (const auto& subtile : tile.sub_tiles)
    CreateJavaTileAndMaybeAddToList(env, jchildren, *subtile.get());

  std::vector<const GURL*> urls;
  for (const ImageMetadata& image : tile.image_metadatas)
    urls.push_back(&image.url);

  return Java_TileConversionBridge_createTileAndMaybeAddToList(
      env, jlist, tile.id, tile.display_text, tile.accessibility_text,
      tile.query_text, urls, tile.search_params, jchildren);
}

ScopedJavaLocalRef<jobject> TileConversionBridge::CreateJavaTiles(
    JNIEnv* env,
    const std::vector<Tile>& tiles) {
  ScopedJavaLocalRef<jobject> jlist = Java_TileConversionBridge_createList(env);

  for (const auto& tile : tiles)
    CreateJavaTileAndMaybeAddToList(env, jlist, tile);

  return jlist;
}

ScopedJavaLocalRef<jobject> TileConversionBridge::CreateJavaTiles(
    JNIEnv* env,
    const std::vector<std::unique_ptr<Tile>>& tiles) {
  ScopedJavaLocalRef<jobject> jlist = Java_TileConversionBridge_createList(env);

  for (const auto& tile : tiles)
    CreateJavaTileAndMaybeAddToList(env, jlist, *tile);

  return jlist;
}

}  // namespace query_tiles
